src/main/kotlin/com/pubnub/api/presence/eventengine/effect/effectprovider/HeartbeatProviderImpl.kt

```diff
@@ -5,7 +5,7 @@ import com.pubnub.api.endpoints.presence.Heartbeat
 import com.pubnub.api.endpoints.remoteaction.RemoteAction
 
 internal class HeartbeatProviderImpl(val pubNub: PubNub) : HeartbeatProvider {
-    override fun getHeartbeatRemoteAction(channels: Set<String>, channelGroups: Set<String>): RemoteAction<Boolean> {
-        return Heartbeat(pubNub, channels.toList(), channelGroups.toList())
+    override fun getHeartbeatRemoteAction(channels: Set<String>, channelGroups: Set<String>, state: Any?): RemoteAction<Boolean> {
+        return Heartbeat(pubNub, channels.toList(), channelGroups.toList(), state)
```

Comment:
Yeah you might be right. At first I thought that because setting state before joining a channel is valid (according to TOTSA), I'll just allow it. But perhaps only the SetState call should set the state once in that case, and only if you join that channel later should Heartbeat also send it. In that case we still need to save it to PresenceData, but filter it before sending Heartbeat.